### PR TITLE
Encode special charaters for vanity URL redirect target

### LINKED ISSUE #26640: Cyrillic URLs encoding issue when configured by Vanity URL Redirect
### Parent Issue

_No response_

### Problem Statement

Cyrillic URL encoding issue, when configured by Vanity URL Redirect to an external Page.

When trying to pass Cyrcllic URL and configuring a vanity URL , The decoded characters are being passed wrongly to the external site

### Steps to Reproduce

1. Create a vanity URL on the demo site (Screenshot attached)
image
2. Try accessing [https://demo.dotcms.com/test/%D0%BC%D0%B5%D1%82%D0%B0%D0%BB%D0%BD%D0%B0-%D0%BA%D0%B0[…]%BD%D0%B5%D0%B7%D0%B5-trender-%D0%B7%D0%B0-apple-watch-4041mm](https://demo.dotcms.com/test/%D0%BC%D0%B5%D1%82%D0%B0%D0%BB%D0%BD%D0%B0-%D0%BA%D0%B0%D0%B8%D1%88%D0%BA%D0%B0-%D1%82%D0%B8%D0%BF-%D0%BC%D0%B8%D0%BB%D0%B0%D0%BD%D0%B5%D0%B7%D0%B5-trender-%D0%B7%D0%B0-apple-watch-4041mm).
3. It is being decoded wrongly as https://www.dotcms.com/destinations/???????-??????-???-????????-trender-??-apple-watch-4041mm
4. This happens not to work if you forward to a page EXTERNALLY,  if you self-reference the site in any way (including using [demo.dotcms.com](http://demo.dotcms.com/)), it appears to work.
5. In the vanity URL configuration, used [dotcms.com](http://dotcms.com/), and you can see it ends up with an improperly encoded version

### Acceptance Criteria

Decoding of write pattern should be sent to the external site

### dotCMS Version

23.01.7
23.10.24

### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

https://dotcms.zendesk.com/agent/tickets/113992
https://dotcms.slack.com/archives/CSHTYUR7H/p1699379538799279

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by dsolistorres**
### Proposed Changes
* Encodes the URL returned as target for a Vanity URL with a 301 redirect. A 301 redirect uses the HTTP Location header to send the target URL back the browser, and it requires the URL to be encoded properly, so the special characters inside the URL are escaped (with %hex-code) 

### Checklist
- [x] Tests

### CODE REVIEW

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java [hunk 1]
```diff
@@ -388,6 +393,62 @@ public boolean handleVanityURLRedirects(final VanityUrlRequestWrapper request, f
...
 
+    /**
+     * Encodes the redirect URL with the parameters from the request.
+     *
+     * @param uri          The URI to redirect to.
+     * @return The encoded redirect URL.
+     */
+    private String encodeRedirectURL(final String uri) {
+        try {
+            boolean hastProtocol = true;
```
Comment (spbolton) [inline dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java:404]: Typo, do you mean hasProtocol

File: dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java [hunk 2]
```diff
@@ -388,6 +393,62 @@ public boolean handleVanityURLRedirects(final VanityUrlRequestWrapper request, f
...
+            String redirectURI = uri;
+            if (uri.startsWith("//")) {
+                hastProtocol = false;
+                redirectURI = "none:" + uri;
+            }
+            final URLUtils.ParsedURL urlToEncode = URLUtils.parseURL(redirectURI);
+            if (urlToEncode == null) {
+                Logger.warn(this, "Could not parse redirect URL: " + uri);
```
Comment (spbolton) [inline dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java:412]: A little concerned we may either be covering up an issue here that should be an exception and will go un noticed, or there is a valid case and we will be logging warnings that will be ignored or clutter the logs?  We should decide whether this is an error or not.

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java (hunk 1 above)
Comment (dsolistorres) [reply dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java]: Typo has been fixed

File: dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java (hunk 2 above)
Comment (dsolistorres) [reply dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java]: Changed to throw exception for an invalid URL

File: dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java [hunk 3]
```diff
@@ -388,6 +393,61 @@ public boolean handleVanityURLRedirects(final VanityUrlRequestWrapper request, f
       return false;
   }
 
+    /**
+     * Encodes the redirect URL with the parameters from the request.
+     *
+     * @param uri          The URI to redirect to.
+     * @return The encoded redirect URL.
+     */
+    private String encodeRedirectURL(final String uri) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java:402]: it seems such as good candidate to expose in an util

**Review by jdotcms — APPROVED** [review]

**Pull request merged**